# Project: cleanning - [candy data]
### this notebook is for educational purpose . in this notebook we will do some cleanning

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">wrangling</a></li>
  
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description 

> i will go through candy data 2017 , this data has :
- Internal_ID
- Q1-Going Out?: Are you actually going trick or treating yourself? "has values `Yes or No`"
- Q2-Gender: It has four different options, `Femal ,Male ,Other ,I'd rather not say`
- Q3-Age: Numerical field
- Q4-Country: Text Field, but users have written their own version of the names. Example, for America, there are entries such as USA, us, US, America so we should consider that while cleanning
- Q5-State/Province: Text Field, but users have written their own version of the names. Same as the country data.
- Q6-Joy Or Despair: All kinds of chocolate bars are the questions with three distinct options to choose from (Joy, Meh, Despair).
- Q7-Joy Other: Text Field to write items not included above that give you JOY. Lots of missing values.
- Q8-Despair Other: Text Field ti write items not included above that give you DESPAIR. Lots of missing values.
- Q9-Other Comments: Text Field. Lots of missing values.
- Q10-Dress: Binary field. Missing values present
- Unnamed: 113
- Q11-Day: Binary Answer Field. Missing values present


# 1- import packages and load data

In [2]:
import pandas as pd
import numpy as np
from fuzzywuzzy import process
from fuzzywuzzy import fuzz


In [3]:
pd.options.mode.chained_assignment = None
df = pd.read_excel("candyhierarchy2017.xlsx", encoding='latin-1')
pd.set_option('display.max_columns', None)

<a id='wrangling'></a>
## Data Wrangling



### General Properties

# 2- show some general properties for the data 

In [4]:
df.head()

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0     90258773            NaN        NaN     NaN         NaN   
1     90272821             No       Male      44        USA    
2     90272829            NaN       Male      49         USA   
3     90272840             No       Male      40          us   
4     90272841             No       Male      23         usa   

  Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                              NaN                NaN   
1                               NM                MEH   
2                         Virginia                NaN   
3                               or                MEH   
4                         exton pa                JOY   

  Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                       
1                                            DESPAIR                                       
2                                                NaN                                       
3                                            DESPAIR                                       
4                                            DESPAIR                                       

  Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
0                           NaN              NaN                      NaN   
1                           JOY              MEH                  DESPAIR   
2                           NaN              NaN                      NaN   
3                           JOY              MEH                      MEH   
4                           JOY          DESPAIR                      MEH   

  Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
0                           NaN              NaN                NaN   
1                       DESPAIR          DESPAIR            DESPAIR   
2                           NaN              NaN                NaN   
3                       DESPAIR              MEH            DESPAIR   
4                       DESPAIR              MEH            DESPAIR   

  Q6 | Broken glow stick Q6 | Butterfinger Q6 | Cadbury Creme Eggs  \
0                    NaN               NaN                     NaN   
1                DESPAIR           DESPAIR                     MEH   
2                    NaN               NaN                     NaN   
3                DESPAIR               MEH                     MEH   
4                DESPAIR               MEH                     MEH   

  Q6 | Candy Corn  \
0             NaN   
1             MEH   
2             NaN   
3         DESPAIR   
4         DESPAIR   

  Q6 | Candy that is clearly just the stuff given out for free at restaurants  \
0                                                NaN                            
1                                            DESPAIR                            
2                                                NaN                            
3                                            DESPAIR                            
4                                            DESPAIR                            

  Q6 | Caramellos Q6 | Cash, or other forms of legal tender Q6 | Chardonnay  \
0             NaN                                       NaN             NaN   
1             MEH                                       JOY             MEH   
2             NaN                                       NaN             NaN   
3             MEH                                       JOY             MEH   
4             JOY                                       MEH             JOY   

  Q6 | Chick-o-Sticks (we don’t know what that is) Q6 | Chiclets  \
0                                              NaN           NaN   
1                                          DESPAIR       DESPAIR   
2                                              NaN           NaN   
3                                              JOY

In [5]:
df.sample(4)


Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE    Q4: COUNTRY  \
1        90272821             No       Male      44           USA    
1113     90277305             No     Female      22             US   
282      90273525             No       Male      48            USA   
2119     90286544             No       Male      49  United States   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
1                                  NM                MEH   
1113                               FL                MEH   
282                             Texas                JOY   
2119                       California                MEH   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                               DESPAIR                                       
1113                                            DESPAIR                                       
282                                             DESPAIR                                       
2119                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
1                              JOY              MEH                  DESPAIR   
1113                           JOY              MEH                  DESPAIR   
282                            JOY          DESPAIR                  DESPAIR   
2119                           JOY              MEH                      MEH   

     Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
1                          DESPAIR          DESPAIR            DESPAIR   
1113                           MEH          DESPAIR            DESPAIR   
282                        DESPAIR          DESPAIR            DESPAIR   
2119                       DESPAIR              MEH            DESPAIR   

     Q6 | Broken glow stick Q6 | Butterfinger Q6 | Cadbury Creme Eggs  \
1                   DESPAIR           DESPAIR                     MEH   
1113                DESPAIR               MEH                     JOY   
282                 DESPAIR               JOY                     JOY   
2119                DESPAIR               MEH                     JOY   

     Q6 | Candy Corn  \
1                MEH   
1113             JOY   
282          DESPAIR   
2119             JOY   

     Q6 | Candy that is clearly just the stuff given out for free at restaurants  \
1                                               DESPAIR                            
1113                                            DESPAIR                            
282                                             DESPAIR                            
2119                                            DESPAIR                            

     Q6 | Caramellos Q6 | Cash, or other forms of legal tender  \
1                MEH                                       JOY   
1113             JOY                                       JOY   
282              MEH                                       JOY   
2119             MEH                                       JOY   

     Q6 | Chardonnay Q6 | Chick-o-Sticks (we don’t know what that is)  \
1                MEH                                          DESPAIR   
1113             JOY                                          DESPAIR   
282              JOY                                          DESPAIR   
2119             JOY                                          DESPAIR   

     Q6 | Chiclets Q6 | Coffee Crisp  \
1          DESPAIR           DESPAIR   
1113       DESPAIR           DESPAIR   
282            MEH           DESPAIR   
2119       DESPAIR               MEH   

     Q6 | Creepy Religious comics/Chick Tracts Q6 | Dental paraphenalia  \
1                                      DESPAIR                  DESPAIR   
1113                                   DESPAIR                  DESPAIR   
282                                    DESPAIR                  DESPAIR   


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Columns: 120 entries, Internal ID to Click Coordinates (x, y)
dtypes: float64(4), int64(1), object(115)
memory usage: 2.3+ MB


In [7]:
df.dtypes

Internal ID                                                                                 int64
Q1: GOING OUT?                                                                             object
Q2: GENDER                                                                                 object
Q3: AGE                                                                                    object
Q4: COUNTRY                                                                                object
Q5: STATE, PROVINCE, COUNTY, ETC                                                           object
Q6 | 100 Grand Bar                                                                         object
Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)     object
Q6 | Any full-sized candy bar                                                              object
Q6 | Black Jacks                                                                           object
Q6 | Bonkers (the ca

In [8]:
df.dtypes.iloc[[*range(7),*range(-12,0)]]

Internal ID                           int64
Q1: GOING OUT?                       object
Q2: GENDER                           object
Q3: AGE                              object
Q4: COUNTRY                          object
Q5: STATE, PROVINCE, COUNTY, ETC     object
Q6 | 100 Grand Bar                   object
Q6 | York Peppermint Patties         object
Q7: JOY OTHER                        object
Q8: DESPAIR OTHER                    object
Q9: OTHER COMMENTS                   object
Q10: DRESS                           object
Unnamed: 113                         object
Q11: DAY                             object
Q12: MEDIA [Daily Dish]             float64
Q12: MEDIA [Science]                float64
Q12: MEDIA [ESPN]                   float64
Q12: MEDIA [Yahoo]                  float64
Click Coordinates (x, y)             object
dtype: object

In [9]:
df.nunique().iloc[[*range(7),*range(-12,0)]]

Internal ID                         2460
Q1: GOING OUT?                         2
Q2: GENDER                             4
Q3: AGE                              106
Q4: COUNTRY                          128
Q5: STATE, PROVINCE, COUNTY, ETC     511
Q6 | 100 Grand Bar                     3
Q6 | York Peppermint Patties           3
Q7: JOY OTHER                        867
Q8: DESPAIR OTHER                    685
Q9: OTHER COMMENTS                   385
Q10: DRESS                             2
Unnamed: 113                           1
Q11: DAY                               2
Q12: MEDIA [Daily Dish]                1
Q12: MEDIA [Science]                   1
Q12: MEDIA [ESPN]                      1
Q12: MEDIA [Yahoo]                     1
Click Coordinates (x, y)             952
dtype: int64

In [10]:
round(df.isnull().sum()/len(df),2)

Internal ID                                                                               0.00
Q1: GOING OUT?                                                                            0.04
Q2: GENDER                                                                                0.02
Q3: AGE                                                                                   0.03
Q4: COUNTRY                                                                               0.03
Q5: STATE, PROVINCE, COUNTY, ETC                                                          0.04
Q6 | 100 Grand Bar                                                                        0.30
Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)    0.30
Q6 | Any full-sized candy bar                                                             0.27
Q6 | Black Jacks                                                                          0.39
Q6 | Bonkers (the candy)                          

In [11]:
df.duplicated().sum()

0

# 3- write summary about what you get form the above functions
    tips : this summary can be about : if there is any missing values , if there a wrong data type , if there a wrong range of values , ....

1- dataframe shape is 2460 rows & 120 columns 
2- columns like Q12, unnamed: 113, Q7, Q8 & Q9 have too many missing values
3- there are too many difference in country spelling resulting too many unique values
4- the main Q is Q6 which contains 103 columns out of 120 in our dataframe
5- no duplicates

# cleaning

### 4- drop all the useless columns and  rows with too much NaN , 
> tip : useless columns can be like `Internal ID`and can be also columns with too much NaN 
- rename columns for easy access

In [13]:
unwanted_columns=['Q5: STATE, PROVINCE, COUNTY, ETC',"Unnamed: 113", "Internal ID", "Click Coordinates (x, y)",
                  'Q7: JOY OTHER', 'Q8: DESPAIR OTHER', 'Q9: OTHER COMMENTS']
df.drop(unwanted_columns, axis = "columns",inplace= True)

In [14]:
df.sample(1)

Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY Q6 | 100 Grand Bar  \
372            NaN       Male      69         USA                MEH   

    Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
372                                                MEH                                       

    Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
372                           JOY              JOY                  DESPAIR   

    Q6 | Bonkers (the board game) Q6 | Bottle Caps Q6 | Box'o'Raisins  \
372                       DESPAIR          DESPAIR                MEH   

    Q6 | Broken glow stick Q6 | Butterfinger Q6 | Cadbury Creme Eggs  \
372                DESPAIR           DESPAIR                     MEH   

    Q6 | Candy Corn  \
372             MEH   

    Q6 | Candy that is clearly just the stuff given out for free at restaurants  \
372                                                MEH                            

    Q6 | Caramellos Q6 | Cash, or other forms of legal tender Q6 | Chardonnay  \
372             JOY                                   DESPAIR             MEH   

    Q6 | Chick-o-Sticks (we don’t know what that is) Q6 | Chiclets  \
372                                          DESPAIR           MEH   

    Q6 | Coffee Crisp Q6 | Creepy Religious comics/Chick Tracts  \
372           DESPAIR                                   DESPAIR   

    Q6 | Dental paraphenalia Q6 | Dots Q6 | Dove Bars Q6 | Fuzzy Peaches  \
372                  DESPAIR       JOY            JOY                MEH   

    Q6 | Generic Brand Acetaminophen Q6 | Glow sticks Q6 | Goo Goo Clusters  \
372                          DESPAIR              JOY                   JOY   

    Q6 | Good N' Plenty Q6 | Gum from baseball cards  \
372                 JOY                      DESPAIR   

    Q6 | Gummy Bears straight up Q6 | Hard Candy Q6 | Healthy Fruit  \
372                          MEH             JOY                MEH   

    Q6 | Heath Bar Q6 | Hershey's Dark Chocolate  \
372        DESPAIR                           JOY   

    Q6 | Hershey’s Milk Chocolate Q6 | Hershey's Kisses  \
372                           JOY                   JOY   

    Q6 | Hugs (actual physical hugs) Q6 | Jolly Rancher (bad flavor)  \
372                          DESPAIR                         DESPAIR   

    Q6 | Jolly Ranchers (good flavor) Q6 | JoyJoy (Mit Iodine!)  \
372                               MEH                       MEH   

    Q6 | Junior Mints Q6 | Senior Mints Q6 | Kale smoothie  \
372               JOY           DESPAIR            DESPAIR   

    Q6 | Kinder Happy Hippo Q6 | Kit Kat Q6 | LaffyTaffy Q6 | LemonHeads  \
372                     MEH      DESPAIR             MEH             JOY   

    Q6 | Licorice (not black) Q6 | Licorice (yes black) Q6 | Lindt Truffle  \
372                       JOY                       JOY                JOY   

    Q6 | Lollipops Q6 | Mars Q6 | Maynards Q6 | Mike and Ike Q6 | Milk Duds  \
372            JOY       JOY           MEH               JOY            JOY   

    Q6 | Milky Way Q6 | Regular M&Ms Q6 | Peanut M&M’s Q6 | Blue M&M's  \
372            JOY               JOY               JOY             JOY   

    Q6 | Red M&M's Q6 | Green Party M&M's Q6 | Independent M&M's  \
372            JOY                    JOY                    JOY   

    Q6 | Abstained from M&M'ing. Q6 | Minibags of chips Q6 | Mint Kisses  \
372                      DESPAIR                    MEH              JOY   

    Q6 | Mint Juleps Q6 | Mr. Goodbar Q6 | Necco Wafers Q6 | Nerds  \
372              JOY              JOY               MEH        JOY   

    Q6 | Nestle Crunch Q6 | Now'n'Laters Q6 | Peeps Q6 | Pencils  \
372                JOY               MEH        MEH      DESPAIR   

    Q6 | Pixy Stix Q6 | Real Housewives of Orange County Season 9 Blue-Ray  \
372            MEH                                                MEH        

    Q6 | Reese’s Peanut Bu

# Renaming columns

In [15]:
df.rename(columns = {'Q1: GOING OUT?' : 'Going Out?',
                          'Q2: GENDER' : 'Gender',
                          'Q3: AGE': 'Age',
                          'Q4: COUNTRY' : 'Country',
                          'Q10: DRESS' : 'Dress',
                          'Q11: DAY': 'Day',
                          'Q12: MEDIA [Daily Dish]' : 'DailyDish',
                          'Q12: MEDIA [Science]': 'Science',
                          'Q12: MEDIA [ESPN]' : 'ESPN',
                          'Q12: MEDIA [Yahoo]': 'Yahoo'}, inplace=True)

In [16]:
df.columns = [x for x in df.columns.str.replace("Q6 \| ",'Candy ')]

In [17]:
df.sample(1)

Going Out? Gender Age Country Candy 100 Grand Bar  \
2085        Yes   Male  49      us                 NaN   

     Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
2085                                                NaN                                        

     Candy Any full-sized candy bar Candy Black Jacks  \
2085                            NaN               NaN   

     Candy Bonkers (the candy) Candy Bonkers (the board game)  \
2085                       NaN                            NaN   

     Candy Bottle Caps Candy Box'o'Raisins Candy Broken glow stick  \
2085               NaN                 NaN                     NaN   

     Candy Butterfinger Candy Cadbury Creme Eggs Candy Candy Corn  \
2085                NaN                      NaN              NaN   

     Candy Candy that is clearly just the stuff given out for free at restaurants  \
2085                                                NaN                             

     Candy Caramellos Candy Cash, or other forms of legal tender  \
2085              NaN                                        NaN   

     Candy Chardonnay Candy Chick-o-Sticks (we don’t know what that is)  \
2085              NaN                                               NaN   

     Candy Chiclets Candy Coffee Crisp  \
2085            NaN                NaN   

     Candy Creepy Religious comics/Chick Tracts Candy Dental paraphenalia  \
2085                                        NaN                       NaN   

     Candy Dots Candy Dove Bars Candy Fuzzy Peaches  \
2085        NaN             NaN                 NaN   

     Candy Generic Brand Acetaminophen Candy Glow sticks  \
2085                               NaN               NaN   

     Candy Goo Goo Clusters Candy Good N' Plenty  \
2085                    NaN                  NaN   

     Candy Gum from baseball cards Candy Gummy Bears straight up  \
2085                           NaN                           NaN   

     Candy Hard Candy Candy Healthy Fruit Candy Heath Bar  \
2085              NaN                 NaN             NaN   

     Candy Hershey's Dark Chocolate Candy Hershey’s Milk Chocolate  \
2085                            NaN                            NaN   

     Candy Hershey's Kisses Candy Hugs (actual physical hugs)  \
2085                    NaN                               NaN   

     Candy Jolly Rancher (bad flavor) Candy Jolly Ranchers (good flavor)  \
2085                              NaN                                NaN   

     Candy JoyJoy (Mit Iodine!) Candy Junior Mints Candy Senior Mints  \
2085                        NaN                NaN                NaN   

     Candy Kale smoothie Candy Kinder Happy Hippo Candy Kit Kat  \
2085                 NaN                      NaN           NaN   

     Candy LaffyTaffy Candy LemonHeads Candy Licorice (not black)  \
2085              NaN              NaN                        NaN   

     Candy Licorice (yes black) Candy Lindt Truffle Candy Lollipops  \
2085                        NaN                 NaN             NaN   

     Candy Mars Candy Maynards Candy Mike and Ike Candy Milk Duds  \
2085        NaN            NaN                NaN             NaN   

     Candy Milky Way Candy Regular M&Ms Candy Peanut M&M’s Candy Blue M&M's  \
2085             NaN                NaN                NaN              NaN   

     Candy Red M&M's Candy Green Party M&M's Candy Independent M&M's  \
2085             NaN                     NaN                     NaN   

     Candy Abstained from M&M'ing. Candy Minibags of chips Candy Mint Kisses  \
2085                           NaN                     NaN               NaN   

     Candy Mint Juleps Candy Mr. Goodbar Candy Necco Wafers Candy Nerds  \
2085               NaN               NaN                NaN         NaN   

     Candy Nestle Crunch Candy Now'n'Laters Candy Peeps Candy Pencils  \
2085                 NaN                NaN         NaN 

# Drop useless rows

In [18]:
df.iloc[:,4:-6].columns

Index(['Candy 100 Grand Bar',
       'Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)',
       'Candy Any full-sized candy bar', 'Candy Black Jacks',
       'Candy Bonkers (the candy)', 'Candy Bonkers (the board game)',
       'Candy Bottle Caps', 'Candy Box'o'Raisins', 'Candy Broken glow stick',
       'Candy Butterfinger',
       ...
       'Candy Three Musketeers', 'Candy Tolberone something or other',
       'Candy Trail Mix', 'Candy Twix',
       'Candy Vials of pure high fructose corn syrup, for main-lining into your vein',
       'Candy Vicodin', 'Candy Whatchamacallit Bars', 'Candy White Bread',
       'Candy Whole Wheat anything', 'Candy York Peppermint Patties'],
      dtype='object', length=103)

In [19]:
df[df.iloc[:,5:-6].isnull().all(axis=1)].index

Int64Index([   0,    2,    6,   10,   18,   19,   24,   25,   30,   35,
            ...
            2437, 2440, 2445, 2446, 2447, 2448, 2449, 2450, 2451, 2458],
           dtype='int64', length=657)

In [20]:
df[df.iloc[:,5:-6].isnull().all(axis=1)]

Going Out?  Gender       Age         Country Candy 100 Grand Bar  \
0           NaN     NaN       NaN             NaN                 NaN   
2           NaN    Male        49             USA                 NaN   
6            No    Male        53             usa                 NaN   
10          Yes    Male        43             NaN                 NaN   
18           No  Female        54   United States                 NaN   
19           No  Female        36   United States                 NaN   
24           No    Male        48   United States                 NaN   
25           No    Male        25  United Kingdom                 NaN   
30           No    Male        48   United States                 NaN   
35           No    Male        16  United States                  NaN   
36          NaN     NaN       NaN             NaN                 NaN   
38           No    Male        33             Usa                 NaN   
40           No    Male        34   United States                 NaN   
42           No    Male        52  United Kingdom                 NaN   
47           No    Male        27             USA                 NaN   
48           No  Female        47             Usa                 NaN   
55          Yes    Male        34   United States                 NaN   
61           No    Male        56          Canada                 NaN   
62           No    Male        43             USA                 NaN   
66           No  Female        42             Usa                 NaN   
67           No  Female        62             USA                 NaN   
76           No    Male        34   United States                 NaN   
78           No    Male        48             USA                 NaN   
84          NaN    Male        33             USA                 NaN   
86           No    Male        37              UK                 NaN   
88           No    Male        31   United States                 NaN   
96           No    Male        64             usa                 NaN   
100         NaN     NaN       NaN             NaN                 NaN   
114          No    Male       NaN             NaN                 NaN   
117          No  Female        44             USA                 NaN   
...         ...     ...       ...             ...                 ...   
2363         No    Male        30         US of A                 NaN   
2364         No  Female  your mom             usa                 NaN   
2369         No    Male        31              US                 NaN   
2374         No  Female        43              US                 NaN   
2386        NaN  Female        17             USA                 NaN   
2388         No  Female        55             USA                 NaN   
2389         No  Female        57              us                 NaN   
2393         No  Female        36             Usa                 NaN   
2394         No    Male        35             USA                 NaN   
2397         No  Female        51             Usa                 NaN   
2401        Yes    Male        49              us                 NaN   
2404         No    Male        41          Canada                 NaN   
2405         No    Male        28   United States                 NaN   
2408         No  Female        51             USA                 NaN   
2409         No  Female        30             USA                 NaN   
2415         No  Female        38          Canada                 NaN   
2416         No    Male        42             USA                 NaN   
2417         No  Female        31          canada                 NaN   
2420         No    Male        45          Canada                 NaN   
2425        Yes  Female        40         Canada                  NaN   
2437         No    Male        25             USA                 NaN   
2440        NaN     NaN       NaN             NaN                 NaN   
2445         No  Female        34          Canada              

In [21]:
df = df.drop(df[df.iloc[:,5:-6].isnull().all(axis=1)].index)

In [22]:
df = df.reset_index()
df.drop(['index'] , axis="columns", inplace=True)

In [23]:
df.head(3)

Going Out? Gender Age Country Candy 100 Grand Bar  \
0         No   Male  44    USA                  MEH   
1         No   Male  40      us                 MEH   
2         No   Male  23     usa                 JOY   

  Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                            DESPAIR                                        
1                                            DESPAIR                                        
2                                            DESPAIR                                        

  Candy Any full-sized candy bar Candy Black Jacks Candy Bonkers (the candy)  \
0                            JOY               MEH                   DESPAIR   
1                            JOY               MEH                       MEH   
2                            JOY           DESPAIR                       MEH   

  Candy Bonkers (the board game) Candy Bottle Caps Candy Box'o'Raisins  \
0                        DESPAIR           DESPAIR             DESPAIR   
1                        DESPAIR               MEH             DESPAIR   
2                        DESPAIR               MEH             DESPAIR   

  Candy Broken glow stick Candy Butterfinger Candy Cadbury Creme Eggs  \
0                 DESPAIR            DESPAIR                      MEH   
1                 DESPAIR                MEH                      MEH   
2                 DESPAIR                MEH                      MEH   

  Candy Candy Corn  \
0              MEH   
1          DESPAIR   
2          DESPAIR   

  Candy Candy that is clearly just the stuff given out for free at restaurants  \
0                                            DESPAIR                             
1                                            DESPAIR                             
2                                            DESPAIR                             

  Candy Caramellos Candy Cash, or other forms of legal tender  \
0              MEH                                        JOY   
1              MEH                                        JOY   
2              JOY                                        MEH   

  Candy Chardonnay Candy Chick-o-Sticks (we don’t know what that is)  \
0              MEH                                           DESPAIR   
1              MEH                                               JOY   
2              JOY                                           DESPAIR   

  Candy Chiclets Candy Coffee Crisp  \
0        DESPAIR            DESPAIR   
1            MEH            DESPAIR   
2        DESPAIR                MEH   

  Candy Creepy Religious comics/Chick Tracts Candy Dental paraphenalia  \
0                                    DESPAIR                   DESPAIR   
1                                    DESPAIR                   DESPAIR   
2                                    DESPAIR                   DESPAIR   

  Candy Dots Candy Dove Bars Candy Fuzzy Peaches  \
0        MEH             JOY             DESPAIR   
1        JOY             JOY                 MEH   
2        MEH             JOY             DESPAIR   

  Candy Generic Brand Acetaminophen Candy Glow sticks Candy Goo Goo Clusters  \
0                           DESPAIR           DESPAIR                DESPAIR   
1                               MEH               JOY                    JOY   
2                           DESPAIR           DESPAIR                    MEH   

  Candy Good N' Plenty Candy Gum from baseball cards  \
0                  MEH                       DESPAIR   
1                  MEH                       DESPAIR   
2                  MEH                       DESPAIR   

  Candy Gummy Bears straight up Candy Hard Candy Candy Healthy Fruit  \
0                           MEH              MEH             DESPAIR   
1                           JOY              MEH                 MEH   
2                           JOY              MEH             DESPAIR   

  Candy Heath Bar Candy Hershey

### 5 - clean `going out` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [24]:
df['Going Out?'].unique()

array(['No', 'Yes', nan], dtype=object)

In [25]:
df['Going Out?'].describe()

count     1739
unique       2
top         No
freq      1500
Name: Going Out?, dtype: object

In [26]:
df['Going Out?'].value_counts()

No     1500
Yes     239
Name: Going Out?, dtype: int64

In [27]:
df['Going Out?'].isnull().sum()

64

In [28]:
df['Going Out?'] = df['Going Out?'].fillna("No")

In [29]:
df['Going Out?'].unique()

array(['No', 'Yes'], dtype=object)

In [30]:
df['Going Out?'].value_counts()

No     1564
Yes     239
Name: Going Out?, dtype: int64

### 6 - clean ` gender` column 
> tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [31]:
df['Gender'].unique()

array(['Male', 'Female', "I'd rather not say", 'Other', nan], dtype=object)

In [32]:
df['Gender'].value_counts()

Male                  1087
Female                 618
I'd rather not say      61
Other                   26
Name: Gender, dtype: int64

In [33]:
df['Gender'].isnull().sum()

11

In [34]:
df['Gender'] = df['Gender'].fillna("I'd rather not say")


In [35]:
df['Gender'] = df['Gender'].str.replace("Other","I'd rather not say")

In [36]:
df['Gender'].value_counts()

Male                  1087
Female                 618
I'd rather not say      98
Name: Gender, dtype: int64

In [37]:
df['Gender'].isnull().sum()

0

In [38]:
df['Gender'].unique()

array(['Male', 'Female', "I'd rather not say"], dtype=object)

### 7 - clean ` age` column 
> tip : note the type of age column , and NaN values
 - try to use `to_numeric` 
 - tip : you should choose which best value to put in place of these NaNs ( mean, median, ....etc)

In [51]:
df['Age'].unique()

array([44. , 40. , 23. ,  nan, 33. , 53. , 56. , 64. , 43. , 37. , 59. ,
       48. , 45. , 49. , 34. , 35. , 38. , 58. , 50. , 47. , 52. , 63. ,
       65. , 41. , 31. , 61. , 46. , 54. , 62. , 29. , 42. , 39. , 36. ,
       32. , 28. , 69. , 67. , 30. , 25. , 22. , 51. , 70. , 24. , 27. ,
       18. , 19. , 57. , 66. , 12. , 60. , 55. , 21. , 11. , 68. , 20. ,
        6. , 10. , 71. , 13. , 26. , 72. ,  9. , 39.4, 74. , 15. ,  8. ,
       75. , 17. ,  7. , 14. , 76. , 16. , 77. ,  4. , 73. ])

In [52]:
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')

In [53]:
df['Age'].unique()

array([44. , 40. , 23. ,  nan, 33. , 53. , 56. , 64. , 43. , 37. , 59. ,
       48. , 45. , 49. , 34. , 35. , 38. , 58. , 50. , 47. , 52. , 63. ,
       65. , 41. , 31. , 61. , 46. , 54. , 62. , 29. , 42. , 39. , 36. ,
       32. , 28. , 69. , 67. , 30. , 25. , 22. , 51. , 70. , 24. , 27. ,
       18. , 19. , 57. , 66. , 12. , 60. , 55. , 21. , 11. , 68. , 20. ,
        6. , 10. , 71. , 13. , 26. , 72. ,  9. , 39.4, 74. , 15. ,  8. ,
       75. , 17. ,  7. , 14. , 76. , 16. , 77. ,  4. , 73. ])

In [54]:
df.Age[df['Age']>80] = np.nan

In [55]:
df['Age'].isnull().sum()

63

In [56]:
Age = df.groupby(['Gender'])['Age'].median()
Age = pd.DataFrame(Age)
Age

Age
Gender                  
Female              40.0
I'd rather not say  39.0
Male                43.0

In [57]:
for i in Age.index:
    df.loc[df['Gender']==i, 'Age'] = df.loc[df['Gender']==i, 'Age'].fillna(Age.loc[i])

In [62]:
df['Age'].isnull().sum()

0

In [60]:
df = df[df['Age'].between(5, 80, inclusive="both")]
df['Age']=df['Age'].astype(int)

In [61]:
df['Age'].unique()

array([44, 40, 23, 33, 53, 56, 64, 43, 37, 59, 48, 45, 49, 34, 35, 38, 58,
       50, 47, 52, 63, 65, 41, 31, 61, 46, 54, 62, 29, 42, 39, 36, 32, 28,
       69, 67, 30, 25, 22, 51, 70, 24, 27, 18, 19, 57, 66, 12, 60, 55, 21,
       11, 68, 20,  6, 10, 71, 13, 26, 72,  9, 74, 15,  8, 75, 17,  7, 14,
       76, 16, 77, 73], dtype=int64)

### 8 - clean ` country` column
> tip : note that country column has country names with different writing ways 
  - use fuzzywuzzy to deal with this problem
  - you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [63]:
df['Country'].unique()


array(['USA ', 'us', 'usa', 'canada', 'Canada', 'Us', 'US', 'Murica',
       'USA', 'uk', 'United States', 'united states', 'Usa',
       'United staes', 'United Kingdom', 'United States of America',
       'UAE', 'United States ', 'England', 'canada ', 'United states',
       'u.s.a.', 'USAUSAUSA', 'america', 'france',
       'United States of America ', 'U.S.A.', 'Mexico', 'Canada ',
       'united states of america', 'US of A', 'UK', 'Unites States',
       'The United States', 'North Carolina ', 'Netherlands', 'Europe',
       'U S', 'u.s.', 'Costa Rica', 'The United States of America',
       'cascadia', 'Greece', 'USA? Hard to tell anymore..', "'merica",
       'Pittsburgh', 'United State', 'France', 'U.S.', 'Can', 'Canae',
       'New York', 'Trumpistan', 'Ireland', 'United Sates', 'Korea',
       'Australia', 'California', 'Japan', 'USa', 'South africa',
       'I pretend to be from Canada, but I am really from the United States.',
       'Usa ', 'Iceland', 'Canada`', 'Scotland

In [64]:
df['Country'].value_counts()

USA                                                                     526
United States                                                           383
usa                                                                     155
Canada                                                                  120
US                                                                       95
Usa                                                                      85
USA                                                                      54
United States of America                                                 45
united states                                                            30
canada                                                                   26
United States                                                            26
us                                                                       24
United states                                                            15
Us          

In [65]:
df['Country'] = df['Country'].str.lower().str.strip()

In [66]:
df['Country'].value_counts()

usa                                                                     822
united states                                                           455
canada                                                                  154
us                                                                      129
united states of america                                                 57
uk                                                                       12
u.s.a.                                                                    8
germany                                                                   7
u.s.                                                                      7
united kingdom                                                            6
america                                                                   5
netherlands                                                               4
japan                                                                     3
france      

In [67]:
us= ['united states', 'merica','us','ussa','u.s.a.','us of a']
for usa in us:
    matches = process.extract(usa, df['Country'], limit = df.shape[0])
    for potential_match in matches:
        if potential_match[1] >= 80:
            df.loc[df['Country'] == potential_match[0],'Country'] = 'usa'

In [68]:
df['Country'].value_counts()

usa                            1509
canada                          154
uk                               12
germany                           7
united kingdom                    6
netherlands                       4
japan                             3
scotland                          3
ireland                           3
france                            3
denmark                           2
switzerland                       2
mexico                            2
canae                             1
cascadia                          1
can                               1
iceland                           1
north carolina                    1
pittsburgh                        1
new york                          1
greece                            1
california                        1
endland                           1
new jersey                        1
canada`                           1
china                             1
europe                            1
trumpistan                  

In [69]:
mapping= {'united kingdom':'uk','murrika':'usa','canada`':'canada','canae':'canada','can':'canada','england':'uk',
          'endland':'uk','north carolina':'usa','california':'usa','new jersey':'usa','cascadia':'usa',
          'new york':'usa','scotland':'uk','ireland':'uk','usa? hard to tell anymore..':'usa'}
df['Country'] = df['Country'].replace(mapping)

In [70]:
df['Country'].value_counts()

usa                     1516
canada                   157
uk                        26
germany                    7
netherlands                4
france                     3
japan                      3
switzerland                2
denmark                    2
mexico                     2
europe                     1
greece                     1
iceland                    1
pittsburgh                 1
china                      1
taiwan                     1
trumpistan                 1
singapore                  1
i don't know anymore       1
uae                        1
korea                      1
narnia                     1
south africa               1
spain                      1
south korea                1
costa rica                 1
Name: Country, dtype: int64

In [71]:
df.Country[~df['Country'].isin(['usa', 'uk','canada'])]='other'

In [72]:
df['Country'].value_counts()

usa       1516
canada     157
other       40
uk          26
Name: Country, dtype: int64

### 9 - clean ` area` column
> tip : area column has the same problem as country columns , but looking at all it's unique values may give us another way to deal with it rather than try to fix it

dropped it because it has many unique values

# 10 - clean Q6 columns¶
        6 has all kinds of chocolate bars and questions with three distinct options to choose from (Joy, Meh, Despair).

In [73]:
df.iloc[:,4:-6].isna().sum().sum()

9343

In [74]:
df.iloc[:,4:-6]=df.iloc[:,4:-6].fillna('MEH') 

In [75]:
df.sample(5)

Going Out?  Gender  Age Country Candy 100 Grand Bar  \
1729         No  Female   57     usa                 JOY   
1386        Yes    Male   26     usa                 JOY   
1437         No    Male   44  canada                 MEH   
1068         No  Female   35  canada                 MEH   
1627         No    Male   15   other                 MEH   

     Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1729                                                MEH                                        
1386                                                MEH                                        
1437                                            DESPAIR                                        
1068                                                MEH                                        
1627                                                MEH                                        

     Candy Any full-sized candy bar Candy Black Jacks  \
1729                            MEH               MEH   
1386                            JOY               MEH   
1437                            JOY               MEH   
1068                            MEH               MEH   
1627                            JOY               MEH   

     Candy Bonkers (the candy) Candy Bonkers (the board game)  \
1729                       MEH                            MEH   
1386                       MEH                            MEH   
1437                       MEH                            MEH   
1068                       JOY                            MEH   
1627                       JOY                        DESPAIR   

     Candy Bottle Caps Candy Box'o'Raisins Candy Broken glow stick  \
1729           DESPAIR             DESPAIR                 DESPAIR   
1386           DESPAIR                 MEH                 DESPAIR   
1437               MEH                 MEH                 DESPAIR   
1068               JOY             DESPAIR                 DESPAIR   
1627           DESPAIR                 JOY                 DESPAIR   

     Candy Butterfinger Candy Cadbury Creme Eggs Candy Candy Corn  \
1729                JOY                      MEH          DESPAIR   
1386                JOY                      JOY              MEH   
1437                JOY                      JOY              MEH   
1068                MEH                      JOY          DESPAIR   
1627                JOY                      MEH              JOY   

     Candy Candy that is clearly just the stuff given out for free at restaurants  \
1729                                            DESPAIR                             
1386                                            DESPAIR                             
1437                                            DESPAIR                             
1068                                                MEH                             
1627                                                MEH                             

     Candy Caramellos Candy Cash, or other forms of legal tender  \
1729          DESPAIR                                        JOY   
1386              JOY                                        JOY   
1437              JOY                                        JOY   
1068              MEH                                        JOY   
1627              JOY                                        JOY   

     Candy Chardonnay Candy Chick-o-Sticks (we don’t know what that is)  \
1729              MEH                                               MEH   
1386          DESPAIR                                           DESPAIR   
1437          DESPAIR                                           DESPAIR   
1068              JOY                                               MEH   
1627          DESPAIR                                           DESPAIR   

     Candy Chiclets Candy Coffee Crisp  \
1729        DESPAIR            DESPAIR   
1386            MEH            DESPAIR   


In [76]:
df.iloc[:,4:-6].isna().sum().sum()

0

### 11 - clean ` dress` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [77]:
df['Dress'].unique()

array(['White and gold', 'Blue and black', nan], dtype=object)

In [78]:
df['Dress']= df['Dress'].fillna('Not interested') 

In [79]:
df['Dress'].unique()

array(['White and gold', 'Blue and black', 'Not interested'], dtype=object)

In [80]:
df['Dress'].isnull().sum()

0

### 12 - clean ` day` column
>tip : you should choose which best value to put in place of these NaNs ( mode, unknown,other ....etc)

In [82]:
df['Day'].unique()

array(['Sunday', 'Friday', nan], dtype=object)

In [83]:
df['Day'].describe()

count       1676
unique         2
top       Friday
freq        1054
Name: Day, dtype: object

In [84]:
df['Day'].isnull().sum()

63

In [85]:
df['Day'] = df['Day'].fillna('Friday')

In [86]:
df['Day'].isnull().sum()

0

In [87]:
df['Day'].value_counts()

Friday    1117
Sunday     622
Name: Day, dtype: int64

# Media cleaning

In [88]:
df.sample(4)

Going Out? Gender  Age Country Candy 100 Grand Bar  \
202         No   Male   48     usa                 JOY   
548         No   Male   21      uk                 MEH   
451         No   Male   32     usa                 JOY   
300         No   Male   35     usa                 JOY   

    Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
202                                            DESPAIR                                        
548                                            DESPAIR                                        
451                                            DESPAIR                                        
300                                                MEH                                        

    Candy Any full-sized candy bar Candy Black Jacks  \
202                            JOY           DESPAIR   
548                            JOY               MEH   
451                            JOY           DESPAIR   
300                            JOY               MEH   

    Candy Bonkers (the candy) Candy Bonkers (the board game)  \
202                   DESPAIR                        DESPAIR   
548                       MEH                            MEH   
451                       MEH                            MEH   
300                       JOY                            MEH   

    Candy Bottle Caps Candy Box'o'Raisins Candy Broken glow stick  \
202           DESPAIR             DESPAIR                 DESPAIR   
548           DESPAIR             DESPAIR                 DESPAIR   
451               JOY                 MEH                 DESPAIR   
300               JOY                 MEH                 DESPAIR   

    Candy Butterfinger Candy Cadbury Creme Eggs Candy Candy Corn  \
202                JOY                      JOY          DESPAIR   
548                MEH                      JOY              JOY   
451                MEH                      MEH          DESPAIR   
300                MEH                      MEH          DESPAIR   

    Candy Candy that is clearly just the stuff given out for free at restaurants  \
202                                            DESPAIR                             
548                                                MEH                             
451                                                MEH                             
300                                            DESPAIR                             

    Candy Caramellos Candy Cash, or other forms of legal tender  \
202              MEH                                        JOY   
548              MEH                                        JOY   
451              MEH                                        JOY   
300              JOY                                        JOY   

    Candy Chardonnay Candy Chick-o-Sticks (we don’t know what that is)  \
202              JOY                                           DESPAIR   
548              MEH                                               MEH   
451              JOY                                               MEH   
300              MEH                                               JOY   

    Candy Chiclets Candy Coffee Crisp  \
202            MEH            DESPAIR   
548            MEH                MEH   
451        DESPAIR                MEH   
300            MEH                JOY   

    Candy Creepy Religious comics/Chick Tracts Candy Dental paraphenalia  \
202                                    DESPAIR                   DESPAIR   
548                                        MEH                   DESPAIR   
451                                    DESPAIR                   DESPAIR   
300                                    DESPAIR                   DESPAIR   

    Candy Dots Candy Dove Bars Candy Fuzzy Peaches  \
202    DESPAIR             JOY             DESPAIR   
548        MEH             MEH                 MEH   
451        JOY             JOY                 MEH   
300    DESPAIR   

In [89]:
media = df.iloc[:,-4:]
for i in media.columns:
    media.loc[media[i] == 1.0, i] = i
df['Media'] = media[media.columns].apply(lambda x: ''.join(x.dropna().astype(str)),axis=1)
df['Media']=df['Media'].astype('category')
df.drop(['DailyDish', 'Science', 'ESPN', 'Yahoo'] , axis="columns", inplace=True)
df.loc[df['Media'] == '', 'Media'] = df['Media'].mode().iloc[0]

In [91]:
df.sample(4)

Going Out? Gender  Age Country Candy 100 Grand Bar  \
354        Yes   Male   40     usa                 JOY   
173         No   Male   52     usa                 MEH   
536         No   Male   39     usa                 MEH   
761         No   Male   32     usa                 JOY   

    Candy Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
354                                            DESPAIR                                        
173                                                JOY                                        
536                                            DESPAIR                                        
761                                            DESPAIR                                        

    Candy Any full-sized candy bar Candy Black Jacks  \
354                            JOY               MEH   
173                            JOY               MEH   
536                            JOY               MEH   
761                            MEH               MEH   

    Candy Bonkers (the candy) Candy Bonkers (the board game)  \
354                       MEH                            MEH   
173                       MEH                            MEH   
536                       MEH                            MEH   
761                       MEH                            MEH   

    Candy Bottle Caps Candy Box'o'Raisins Candy Broken glow stick  \
354               JOY                 MEH                     MEH   
173               JOY             DESPAIR                 DESPAIR   
536           DESPAIR                 MEH                 DESPAIR   
761               MEH             DESPAIR                 DESPAIR   

    Candy Butterfinger Candy Cadbury Creme Eggs Candy Candy Corn  \
354                JOY                      JOY              JOY   
173                JOY                      JOY              MEH   
536                JOY                      JOY              JOY   
761                MEH                  DESPAIR          DESPAIR   

    Candy Candy that is clearly just the stuff given out for free at restaurants  \
354                                            DESPAIR                             
173                                            DESPAIR                             
536                                                MEH                             
761                                            DESPAIR                             

    Candy Caramellos Candy Cash, or other forms of legal tender  \
354              JOY                                        JOY   
173              JOY                                        JOY   
536              JOY                                        JOY   
761              JOY                                        MEH   

    Candy Chardonnay Candy Chick-o-Sticks (we don’t know what that is)  \
354              JOY                                               JOY   
173          DESPAIR                                           DESPAIR   
536              MEH                                           DESPAIR   
761          DESPAIR                                           DESPAIR   

    Candy Chiclets Candy Coffee Crisp  \
354        DESPAIR            DESPAIR   
173        DESPAIR            DESPAIR   
536        DESPAIR            DESPAIR   
761        DESPAIR            DESPAIR   

    Candy Creepy Religious comics/Chick Tracts Candy Dental paraphenalia  \
354                                    DESPAIR                   DESPAIR   
173                                    DESPAIR                   DESPAIR   
536                                    DESPAIR                   DESPAIR   
761                                        MEH                   DESPAIR   

    Candy Dots Candy Dove Bars Candy Fuzzy Peaches  \
354        MEH             JOY             DESPAIR   
173    DESPAIR             JOY                 MEH   
536    DESPAIR             JOY                 MEH   
761        JOY   

### 13 - check
> after finish cleaning you should make sure your data is clean and that :
- there is no wrong data type
- there is no NaN values 
- all the column are cleaing and ready for the next step `EDA`

### 14 - write a summary about all the cleaning steps 

1- understanding data and detect the important columns 
2- started with dropping the columns with many null values which is difficult to clean (7 columns)
3- renaming the columns to be easier to handle
4- dropping the garbage rows to make data more accurate (nearly 700 row)
5- started to fill the nan values in all columns using statistical ways
